cmd/cosign/cli/verify/verify_attestation.go

```diff
@@ -162,22 +163,58 @@ func (c *VerifyAttestationCommand) Exec(ctx context.Context, images []string) (e
 		}
 
 		var validationErrors []error
-		for _, vp := range verified {
-			var payloadData map[string]interface{}
+		_, validationErrors, err := DecodeAttestationPayload(verified, c.PredicateType, cuePolicies, regoPolicies, validationErrors)
+		if err != nil {
+			return fmt.Errorf("error Decoding and validating attestation payloads %v", err)
+		}
 
-			p, err := vp.Payload()
-			if err != nil {
-				return errors.Wrap(err, "could not get payload")
+		if len(validationErrors) > 0 {
+			fmt.Fprintf(os.Stderr, "There are %d number of errors occurred during the validation:\n", len(validationErrors))
+			for _, v := range validationErrors {
+				_, _ = fmt.Fprintf(os.Stderr, "- %v\n", v)
 			}
+			return fmt.Errorf("%d validation errors occurred", len(validationErrors))
+		}
 
-			err = json.Unmarshal(p, &payloadData)
+		// TODO: add CUE validation report to `PrintVerificationHeader`.
+		PrintVerificationHeader(imageRef, co, bundleVerified)
+		_ = bundleVerified
+		if c.DecodePayload {
+			decodedPayload, _, err := DecodeAttestationPayload(verified, "", []string{}, []string{}, []error{})
 			if err != nil {
-				return errors.Wrap(err, "unmarshal payload data")
+				return fmt.Errorf("error Decoding Attestation Payload: %v", err)
 			}
+			PrintDecodedAttestation(decodedPayload)
+		} else {
+			// The attestations are always JSON, so use the raw "text" mode for outputting them instead of conversion
+			PrintVerification(imageRef, verified, "text")
+		}
 
-			predicateURI, ok := options.PredicateTypeMap[c.PredicateType]
+	}
+
+	return nil
+}
+
+func DecodeAttestationPayload(verified []oci.Signature, predicateType string, cuePolicies, regoPolicies []string, validationErrors []error) ([]byte, []error, error) {
```

Comment:
I don't think the policies belong in this function. I would expect `DecodeAttestationPayload` to do just that.

Reply:
I think this should take one `oci.Signature`, rather than a slice. If there are multiple signatures then you would expect multiple payloads. It looks to me like this is just going to return the last payload it happens to find.

Reply:
In fact, maybe it should just be a `[]byte` (the payload)?

Reply:
yeah completely agree, just couldn't quite figure out how best to split things up without it being really clunky. Will take a look it later and give it another stab.

Reply:
> I think this should take one oci.Signature, rather than a slice. If there are multiple signatures then you would expect multiple payloads. It looks to me like this is just going to return the last payload it happens to find.

aha! You're right! Thank you for spotting. I was thinking this earlier but had to pivot to other work. Will sort this out in due course